src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1439,23 +1429,125 @@ public static IEnumerable<int> Range(int start, int count)
         {
-            for (int end = start + count; start != end; start++) yield return start;
+            private readonly int _start;
+            private readonly int _end;
+            
+            public RangeIterator(int start, int count)
+            {
+                _start = start;
+                _end = start + count;
+            }
+            
+            public override Iterator<int> Clone()
+            {
+                return new RangeIterator(_start, _end - _start);
+            }
+            
+            public override bool MoveNext()
+            {
+                switch(state)
+                {
+                    case 1:
+                        if (_start == _end) break;
+                        current = _start;
+                        state = 2;
+                        return true;
+                    case 2:
+                        if (++current == _end) break;
+                        return true;
+                }
+                state = -1;
+                return false;
+            }
+
+            public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector)
+            {
+                return new WhereSelectEnumerableIterator<int, TResult>(this, null, selector);
+            }
+
+            public override IEnumerable<int> Where(Func<int, bool> predicate)
+            {
+                return new WhereEnumerableIterator<int>(this, predicate);
+            }
+
+            public int[] ToArray()
+            {
+                int[] array = new int[_end - _start];
+                int cur = _start;
+                for (int i = 0; i != array.Length; ++i)
+                {
+                    array[i] = cur;
+                    ++cur;
+                }
+
+                return array;
+            }
         }
 
         public static IEnumerable<TResult> Repeat<TResult>(TResult element, int count)
         {
             if (count < 0) throw Error.ArgumentOutOfRange("count");
-            return RepeatIterator<TResult>(element, count);
+            return new RepeatIterator<TResult>(element, count);
         }
 
-        private static IEnumerable<TResult> RepeatIterator<TResult>(TResult element, int count)
+        private sealed class RepeatIterator<TResult> : Iterator<TResult>, IArrayProvider<TResult>
         {
-            for (int i = 0; i < count; i++) yield return element;
+            private readonly int _count;
+            private int _sent;
+
+            public RepeatIterator(TResult element, int count)
+            {
+                current = element;
+                _count = count;
+            }
+            
+            public override Iterator<TResult> Clone()
+            {
+                return new RepeatIterator<TResult>(current, _count);
+            }
+            
+            public override void Dispose()
+            {
+                // Don't let base Dispose wipe current.
+                state = -1;
+            }
```

Comment:
RE: clearing of Current in Dispose.

I think the behavior of Current after Disposing is undefined. Some implementations throw ObjectDisposed, some return last value, some return default value ...

Maintaining same behavior of "Current after Dispose" does not seem very important, just because behavior is undefined in spec in inconsistent among implementations.

The reason for clearing Current in dispose is typically an opportunistic optimization to not keep it rooted longer than necessary. So IMO clearing is a mildly good thing (unless we know T is a primitive value), but not extremely important. 

It is a bit more important for iterators that are also enumerables. - While it is not common for disposed iterators to be retained, the enumerable itself can be kept alive for quite a long time and if that keeps an instance of Current alive after the first iterator (same instance as enumerable) is disposed, it may become a cause of unexpected "leaks". 

It may make sense to review all the Dispose implementations here and make sure we clear Current where convenient, but again, I do not see this as highly important.